NangoHQ/nango
# feat(sync): allow to force a full sync anytime

## Context

Fixes NAN-305

Customers want to trigger a full resync without having to delete a sync / connection. We assume they want to get a fresh data set and they might not have `track_deletes` activated. 

## Proposal

There were no proper mechanism in place to achieve that, so I had to modify a lot of logic here and there to make it work.
- `GET /sync/trigger { full_resync: true }`
- Cancel running schedule
- Delete all the records associated with the SyncId
- Clear the lastSyncDate so the scripts can start from scratch
- Trigger an initial Sync

This operation is destructive and if the sync fails the customers will have no data. But it's the intended behavior and not different from what they were doing manually. We can improve this behavior later.

## Describe your changes

- Orchestrator `runSyncCommand` has a different execution path when `fullResync` is passed.
We use the initial sync as a matter of clarity and logging. We don't want to have "Continuous" when it's in fact a "from scratch sync". I also didn't wanted to have a new "workflow" because of concurrency issues; initial sync concurrency is already handled so it limits the logic required to make this works.


- Separate Initial Sync and Init activities into two distinct Activity.  
Currently we were passing `activitLogId` to temporal to have a continuous log. But we ended up storing this `activityLogId` for ever and could lead to mistakes (e.g: writing to the wrong activity log).
In this case, we were logging to the first very Activity which was incorrect. I had to improvise a bit here so let me know what you think.

<img width="1487" alt="Screenshot 2024-02-19 at 11 55 32" src="https://github.com/NangoHQ/nango/assets/1637651/883eb775-6807-42b5-b461-dafd0ad85c16">




- Add missing `SYNC_CANCEL` in analytics
Unrelated but I saw it missing


## How to test this?

Use this integrations

```ts
import type { NangoSync } from './models';

export default async function fetchData(nango: NangoSync): Promise<void> {
    if (nango.lastSyncDate) {
        await nango.log('With sync date');
        await nango.batchSave(
            [
                { id: 1, created_at: new Date() },
                { id: 2, created_at: new Date() }
            ],
            'Simple'
        );
    } else {
        await nango.log('No sync date');
        await nango.batchSave([{ id: 1, created_at: new Date() }], 'Simple');
    }

    await nango.log('✅ end');
}
```

- Create this integrations
- It will create 1 record at the first execution, then 2 at each subsequent executions.
- Call /sync/trigger with `full_resync: false` multiple times
- Check DB `_nango_sync_date_records`: you should have 2 records
- Call /sync/trigger with `full_resync: true`
- Check DB `_nango_sync_date_records`: you should have 1 record
